askama_shared/src/generator.rs

```diff
@@ -1523,51 +1567,73 @@ impl Buffer {
     }
 }
 
+// type SetChain<'a, T> = MapChain<'a, T, ()>;
+
 #[derive(Debug)]
-struct SetChain<'a, T: 'a>
+struct MapChain<'a, K: 'a, V: 'a>
 where
-    T: cmp::Eq + hash::Hash,
+    K: cmp::Eq + hash::Hash,
 {
-    parent: Option<&'a SetChain<'a, T>>,
-    scopes: Vec<HashSet<T>>,
+    parent: Option<&'a MapChain<'a, K, V>>,
+    scopes: Vec<HashMap<K, V>>,
 }
 
-impl<'a, T: 'a> SetChain<'a, T>
+impl<'a, K: 'a, V: 'a> MapChain<'a, K, V>
 where
-    T: cmp::Eq + hash::Hash,
+    K: cmp::Eq + hash::Hash,
 {
-    fn new() -> SetChain<'a, T> {
-        SetChain {
+    fn new() -> MapChain<'a, K, V> {
+        MapChain {
             parent: None,
-            scopes: vec![HashSet::new()],
+            scopes: vec![HashMap::new()],
         }
     }
 
-    fn with_parent<'p>(parent: &'p SetChain<T>) -> SetChain<'p, T> {
-        SetChain {
+    fn with_parent<'p>(parent: &'p MapChain<K, V>) -> MapChain<'p, K, V> {
+        MapChain {
             parent: Some(parent),
-            scopes: vec![HashSet::new()],
+            scopes: vec![HashMap::new()],
         }
     }
 
-    fn contains(&self, val: T) -> bool {
-        self.scopes.iter().rev().any(|set| set.contains(&val))
+    fn contains(&self, key: &K) -> bool {
+        self.scopes.iter().rev().any(|set| set.contains_key(&key))
             || match self.parent {
-                Some(set) => set.contains(val),
+                Some(set) => set.contains(key),
                 None => false,
             }
     }
 
+    fn get(&self, key: &K) -> Option<&V> {
+        self.get_skip(key, 0)
+    }
+
+    /// Returns `Some` if `key` exists, while `skip` represents
+    /// the amount of previous scopes to skip in reverse.
+    fn get_skip(&self, key: &K, skip: usize) -> Option<&V> {
+        self.scopes
+            .iter()
+            .rev()
+            .skip(skip)
+            .filter_map(|set| set.get(&key))
+            .next()
+            .or_else(|| match self.parent {
+                Some(set) => set.get(key),
+                None => None,
+            })
+    }
+
     fn is_current_empty(&self) -> bool {
         self.scopes.last().unwrap().is_empty()
     }
 
-    fn insert(&mut self, val: T) {
-        self.scopes.last_mut().unwrap().insert(val);
+    fn insert(&mut self, key: K, val: V) {
+        let old_val = self.scopes.last_mut().unwrap().insert(key, val);
```

Comment:
Maybe add a comment documenting why the asserted invariant should hold?

Reply:
Hmm, we could also remove the `assert` because Rust itself will error when attempting to compile the generated template. Basically, the cases are e.g. `{% macro f(a, a) %}` and `{% let (a, a) = ... %}` which both results in a compile error along the lines of "identifier `a` used more than once" .

I'll remove it. Because if we want to add `"identifier {:?} used more than once"` to the assert, then `insert` requires `where K: Debug + Clone/Copy`.